blog/2024-11-mastering-effect-and-afterRenderEffect/README.md

```diff
@@ -0,0 +1,477 @@
+
+Let’s dive into the specifics! 🚀
+
+
+
+## Core Differences Between `effect()` and `afterRenderEffect()`
+
+Both `effect()` and `afterRenderEffect()` are designed to track and respond to changes in signals, but they differ in timing and use cases.
+
+- **`effect()`** runs as part of the Angular change detection cycle and can now safely modify signals without any additional flags.
+- **`afterRenderEffect()`** is a lower-level API that executes after the DOM has been updated. 
+  It’s particularly suited for tasks that require interacting directly with the DOM, such as measuring element sizes or making complex visual updates in phases.
+
+Here’s a simple comparison to illustrate how these functions operate:
+
+```typescript
+counter = signal(0);
+
+effect(() => {
+  console.log(`Current counter value: ${this.counter()}`);
+});
+
+afterRenderEffect(() => {
+  console.log('DOM rendering completed for this component');
+});
+```
+
+As expected, the console output for `afterRenderEffect` is triggered after the output of `effect`.
+
+
+## Introducing `effect()`
+
+Angular supports two types of effects: **component effects** and **root effects**. 
+Component effects are initiated when `effect()` is called within a component, directive, or a service tied to them. 
+Root effects, on the other hand, are initiated when `effect()` is called outside the component tree, such as in a root service, or by setting the `forceRoot` option.
+
+The main difference between these effect types is their timing. 
+Component effects operate as part of Angular’s change detection, which allows them to safely read input signals and manage views dependent on component state.
+Root effects, however, run as microtasks, independent of the component tree or change detection.
+
+In this article, we only discuss **component effects**.
+
+
+### Example for `effect()`: setting multipe things at once
+
+Consider the example below, where `effect()` is used to synchronize form fields based on the input signal `currentBook`.
+The API for reactive forms has not been updated to work hand in hand with signals, so we still need to patch the form as we have done in the past.
+However, some improvements to the Reactive Forms API have already been promised.
+We also want to set another signal after we have patched the form. 
+
+Here is our example of a form that can create a new book and edit an existing book:
+
+```typescript
+@Component({
+  selector: 'app-book-form',
+  imports: [ReactiveFormsModule],
+  template: `
+    @let c = bookForm.controls;
+
+    <form [formGroup]="bookForm" (ngSubmit)="submitForm()">
+      <input [formControl]="c.isbn" placeholder="ISBN" />
+      <input [formControl]="c.title" placeholder="Title" />
+      <textarea [formControl]="c.description" placeholder="Description"></textarea>
+      <button type="submit">
+        @if (isEditMode()) { Edit } @else { Create }
+      </button>
+    </form>
+  `,
```

Comment:
@d-koppenhagen Is a placeholder enough for good accessibility? Feels like it's not and we should improve it if necessary.

Reply:
Yes, we do. I don't mean that we should bloat the example with everything possible, but just use something different than placeholder. I already hear people saying "But you also use placeholders in your example, why shouldn't I do this?"

Let's see how this can be made better (if necessary) and I suspect that it won't be much longer.